In [1]:
from sklearn.metrics import classification_report, confusion_matrix


Step 1: Evaluation Tools
We are importing tools to evaluate the performance of each model.  
- `classification_report()` shows precision, recall, f1-score, and accuracy.  
- `confusion_matrix()` shows true/false positives and negatives.

In [7]:
import pandas as pd

df = pd.read_csv("C:/Users/user/Desktop/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [8]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(df['TotalCharges'].median())

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for column in df.columns:
    if df[column].dtype == 'object':
        df[column] = le.fit_transform(df[column])


In [9]:
X = df.drop('Churn', axis=1)
y = df['Churn']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [10]:
# Split the dataset again 
X = df.drop('Churn', axis=1)
y = df['Churn']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [12]:
from sklearn.linear_model import LogisticRegression

log_model = LogisticRegression(max_iter=2000)
log_model.fit(X_train, y_train)

y_pred_log = log_model.predict(X_test)


C:\Users\user\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


##  Logistic Regression Evaluation
We use the trained logistic regression model to:
- Predict on test data (`y_pred_log`)
- Compare with actual test labels (`y_test`)
- View classification report and confusion matrix

In [25]:
from sklearn.metrics import classification_report, confusion_matrix

# Logistic Regression
print(" Logistic Regression:\n")
print(classification_report(y_test, y_pred_log))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_log))


 Logistic Regression:

              precision    recall  f1-score   support

           0       0.85      0.90      0.88      1036
           1       0.68      0.57      0.62       373

    accuracy                           0.81      1409
   macro avg       0.77      0.74      0.75      1409
weighted avg       0.81      0.81      0.81      1409

Confusion Matrix:
 [[935 101]
 [160 213]]


In [26]:
# Logistic Regression Evaluation
print(" Logistic Regression Results:\n")
print(classification_report(y_test, y_pred_log))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_log))


 Logistic Regression Results:

              precision    recall  f1-score   support

           0       0.85      0.90      0.88      1036
           1       0.68      0.57      0.62       373

    accuracy                           0.81      1409
   macro avg       0.77      0.74      0.75      1409
weighted avg       0.81      0.81      0.81      1409

Confusion Matrix:
 [[935 101]
 [160 213]]


In [20]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)


##  Random Forest Evaluation
Same as above — now we evaluate the predictions from the Random Forest model.
We will compare it against Logistic Regression to see if performance improves.

In [27]:
from sklearn.metrics import classification_report, confusion_matrix

print("\n Random Forest Results:\n")
print(classification_report(y_test, y_pred_rf))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_rf))



 Random Forest Results:

              precision    recall  f1-score   support

           0       0.83      0.90      0.86      1036
           1       0.64      0.49      0.56       373

    accuracy                           0.79      1409
   macro avg       0.74      0.70      0.71      1409
weighted avg       0.78      0.79      0.78      1409

Confusion Matrix:
 [[933 103]
 [189 184]]


In [23]:
from xgboost import XGBClassifier

xgb_model = XGBClassifier(use_label_encoder=False, eval_metric='logloss')
xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)


C:\Users\user\anaconda3\Lib\site-packages\xgboost\training.py:183: UserWarning: [12:33:06] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


##  XGBoost Evaluation
XGBoost is often powerful for classification.  
We check if it gives better recall and f1-score for 'churn = 1' class.


In [28]:
print("\n XGBoost Results:\n")
print(classification_report(y_test, y_pred_xgb))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_xgb))



 XGBoost Results:

              precision    recall  f1-score   support

           0       0.84      0.88      0.86      1036
           1       0.61      0.53      0.57       373

    accuracy                           0.78      1409
   macro avg       0.72      0.70      0.71      1409
weighted avg       0.78      0.78      0.78      1409

Confusion Matrix:
 [[907 129]
 [175 198]]


## 📈 Evaluation Summary

| Model              | Accuracy | Precision (Churn=1) | Recall (Churn=1) | F1-Score (Churn=1) |
|-------------------|----------|---------------------|------------------|--------------------|
| Logistic Regression |   0.81         0.68                  0.57       |         0.62
| Random Forest       |   0.79         0.64                  0.49                 0.56          |
| XGBoost                 0.78           0.61                  0.53               0.57               |

 Based on the above:
- XGBoost gave the best performance.
- Since our goal is to detect churn, we prefer the model with highest recall for churn class (1).

 Final Chosen Model: XGBoost
